pkg/admission/certoutput/certoutput.go

```diff
@@ -0,0 +1,374 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certoutput
+
+import (
+	"bytes"
+	"fmt"
+	"net/url"
+	"strings"
+
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	apitypes "k8s.io/apimachinery/pkg/types"
+
+	"github.com/kubernetes-sigs/controller-runtime/pkg/admission/certinput"
+	"github.com/kubernetes-sigs/controller-runtime/pkg/client"
+)
+
+const (
+	// SecretCertProvisionAnnotationKeyPrefix should be used in an annotation in the following format:
+	// secret.certprovisioner.kubernetes.io/<webhook-name>: <secret-namespace>/<secret-name>
+	// the webhook cert manager library will provision the certificate for the webhook by
+	// storing it in the specified secret.
+	SecretCertProvisionAnnotationKeyPrefix = "secret.certprovisioner.kubernetes.io/"
+
+	// FSCertProvisionAnnotationKeyPrefix should be used in an annotation in the following format:
+	// fs.certprovisioner.kubernetes.io/<webhook-name>: path/to/certs/
+	// the webhook cert manager library will provision the certificate for the webhook by
+	// storing it under the specified path.
+	// format: local.certprovisioner.kubernetes.io/webhookName: path/to/certs/
+	FSCertProvisionAnnotationKeyPrefix = "fs.certprovisioner.kubernetes.io/"
+
+	// CACertName is the name of the CA certificate
+	CACertName = "ca-cert.pem"
+	// ServerKeyName is the name of the server private key
+	ServerKeyName = "key.pem"
+	// ServerCertName is the name of the serving certificate
+	ServerCertName = "cert.pem"
+)
+
+// Options are options for configure a CertOutput.
+type Options struct {
+	Type *Type
+}
+
+const (
+	// SecretType is a type of CertsOutput that will write certificates to secrets.
+	SecretType Type = "secret-certoutput-type"
+)
+
+// Type is the type of the CertsOutput.
+type Type string
+
+// CertsOutput provides method to handle webhooks.
+type CertsOutput interface {
+	// Handle processes one webhook and ensure it has the certificate in the right place.
+	Handle(webhook *admissionregistrationv1beta1.Webhook) error
+}
+
+// New builds a new CertsOutput instance for the provided webhook configuration.
+func New(webhookConfig runtime.Object, client client.Client, cp certinput.CertInput, op Options) (CertsOutput, error) {
```

Comment:
For client, WDYT of injecting it?

Reply:
What will it look like? Some pseudo code will be helpful.